### PR TITLE
[ENH] Add Time Series Transformer (Timer) for Forecasting

### PR DESCRIPTION
**Pull request by Sohaib-Ahmed21**
Added timer forecasting model to sktime with comments, examples and end to end tested code.

#### Reference Issues/PRs

Fixes: #7700 

#### What does this implement/fix? Explain your changes.

Added timer forecasting model to sktime with comments, examples and end to end tested code.

#### What should a reviewer concentrate their feedback on?

There is no particular thing to focus more upon, general review will be helpful.
#### Did you add any tests for the change?

Added tests to get_test_params() to check multiple scenarios.

##### For all contributions
- [] I've added myself to the [list of contributors](https://github.com/sktime/sktime/blob/main/CONTRIBUTORS.md) with any new badges I've earned :-)
  How to: add yourself to the [all-contributors file](https://github.com/sktime/sktime/blob/main/.all-contributorsrc) in the `sktime` root directory (not the `CONTRIBUTORS.md`). Common badges: `code` - fixing a bug, or adding code logic. `doc` - writing or improving documentation or docstrings. `bug` - reporting or diagnosing a bug (get this plus `code` if you also fixed the bug in the PR).`maintenance` - CI, test framework, release.
  See here for [full badge reference](https://allcontributors.org/docs/en/emoji-key)
- [x] Optionally, for added estimators: I've added myself and possibly to the `maintainers` tag - do this if you want to become the owner or maintainer of an estimator you added.
  See here for further details on the [algorithm maintainer role](https://www.sktime.net/en/latest/get_involved/governance.html#algorithm-maintainers).
- [x] The PR title starts with either [ENH], [MNT], [DOC], or [BUG]. [BUG] - bugfix, [MNT] - CI, test framework, [ENH] - adding or improving code, [DOC] - writing or improving documentation or docstrings.

##### For new estimators
- [ ] I've added the estimator to the API reference - in `docs/source/api_reference/taskname.rst`, follow the pattern.
- [x] I've added one or more illustrative usage examples to the docstring, in a pydocstyle compliant `Examples` section.
- [x] If the estimator relies on a soft dependency, I've set the `python_dependencies` tag and ensured
  dependency isolation, see the [estimator dependencies guide](https://www.sktime.net/en/latest/developer_guide/dependencies.html#adding-a-soft-dependency).

### CODE REVIEW

**Comment by Sohaib-Ahmed21** [comment]
### Explanation of `TimerForecaster`

The `TimerForecaster` is designed to perform **Zero-Shot Forecasting** for univariate time series, leveraging Transformer-based models. It works by utilizing a pre-trained model (e.g., from HuggingFace) to predict future values for time series data. Below is a breakdown of how it works:

1. **`trust_remote_code`**:
   - This parameter determines whether custom models from HuggingFace repositories are allowed to run. By setting it to `True`, the code will trust and execute models that have custom components defined in their repository.
   - **Security Warning**: Use this only if you trust the source of the model as it could potentially run arbitrary code.
   - To use `TimerForecaster`, set this param to **True**.

2. **Dataset Creation**:
   - The `fit` method takes a time series dataset and processes it for model loading and finetuning. For fitting, hugging face trainer is used and data is loaded from custom made `PytorchDataset`. 

3. **`fit_strategy`**:
   - This controls how the model is fine-tuned or used for forecasting:
     - **`zero-shot`**: The model performs predictions without further training, using pre-trained weights.
     - **`minimal`**: Default value, only a small subset of the model(mismatched or missing layers compared to pretrained one) is fine-tuned, which saves resources and is useful for quick adaptation.
     - **`full`**: Fine-tunes the entire model on the provided data, ensuring the best performance but requiring more computational power.

4. **Forecasting**:
   - After fitting, the `predict()` function generates future predictions based on the provided `forecasting_horizon`. The model supports global forecasting and predictions are based on the model's understanding of the time series data and can be adjusted based on the selected fine-tuning strategy.

The approach is similar to `TTM` (Temporal Transformer Models), where the core model relies on pre-trained weights, but flexibility is provided with different fine-tuning options(later to be included in ttm through https://github.com/sktime/sktime/pull/7648) depending on computational resources and use case.

**Review by fkiraly — CHANGES_REQUESTED** [review]
Nice! Before I review more deeply, may I request an explanation how this differs from the existing transformers/hf interfaces? Could it be handled by an extension of the current hf interface? What were your thoughts on that?

FYI @geetu040, @benHeid

**Comment by Sohaib-Ahmed21** [comment]
> Nice! Before I review more deeply, may I request an explanation how this differs from the existing transformers/hf interfaces? Could it be handled by an extension of the current hf interface? What were your thoughts on that?


The architecture of the timer model(its config) , loading(config&model) and generating methods and inputs and outputs format are different from those models currently there in existing hftransformers interface which all have more or less same things(methods,formats,config) mentioned above. 

But keeping in view the redesign, I believe that 3rd party foundation models should be integrated all under one interface. To get started, maybe I should start integrating this model in existing interface and then we proceed on integrating other 3rd party foundation models based on the outcomes. Input from @geetu040  and @benHeid will also be helpful.

**Comment by geetu040** [comment]
> Nice! Before I review more deeply, may I request an explanation how this differs from the existing transformers/hf interfaces? Could it be handled by an extension of the current hf interface? What were your thoughts on that?

@fkiraly Looking at the code base, I think it should not be that difficult to integrate with the existing `HFTransformerForecaster`. It might be a little challenging but definitely worth the try and I really believe its technically sound and can be done.

> To get started, maybe I should start integrating this model in existing interface and then we proceed on integrating other 3rd party foundation models based on the outcomes

@Sohaib-Ahmed21, yes we have discussed this previously in sktime, that we need to bring these similar sounding models under the same hood, that will be quite some work and a design challenge but bringing all the huggingface transformers under a unified interface will be really nice!

---

I am not really sure how the updated `HFTransformerForecaster` or a new `HFTransformerAdapter` looks, and I have not really thought much of the design, but I'll think about this and give my ideas as well. Now if you want to give `Timer` a try to be extended with the `HFTransformerForecaster`, I would really espouse the idea. Good Luck with that! and feel free to ask any question or suggest any design for review.

**Comment by geetu040** [comment]
From little that I can think of:
Following segments can be combined for different models
- all fit_strategies should work for any model architecute
- config and model should be loaded using AutoConfig and AutoModel for all architectures
- same dataset class should be reused by all models

Following segments might be different for each model
- we do make some changes in the `config`, which I think will be different for each architecture
- extra steps in data processing (like padding, masking) can be different for architectures, though they can be supported in a single dataset class
- `predict` function can be different for each architecutre, because some use `PredictionHead`, while some use `CausalHead`, and some generate predictions through logits and some use `.generate()` for better and controllable autoregressive outputs.

These are just some drafted pointed, it may require more research and thinking ....

**Comment by Sohaib-Ahmed21** [comment]
> I am not really sure how the updated `HFTransformerForecaster` or a new `HFTransformerAdapter` looks, and I have not really thought much of the design, but I'll think about this and give my ideas as well. Now if you want to give `Timer` a try to be extended with the `HFTransformerForecaster`, I would really espouse the idea. Good Luck with that! and feel free to ask any question or suggest any design for review.
 
Thanks for your input  @geetu040 . I will think of the design approach and then start integrating **timer** into existing **hftransformers**.  Then hopefully,  throughout the process, new things and efficient approaches will be discovered and then we can move to integrating other models into **hftransformers** as well.

**Comment by Sohaib-Ahmed21** [comment]
@fkiraly @geetu040 @benHeid Now I have Integrated **Timer** into `HFTransformersForecaster` and laid down the basis using utils for adding other 3rd party foundation models having similar interface.

I have also added Global Forecasting Support for `HFTransformersForecaster` as mentioned in #6673 .
Your review on current design will be helpful. Thanks!

**Comment by Sohaib-Ahmed21** [comment]
@geetu040 @fkiraly if we're not moving towards HFTransformersForecaster redesign, then can this pull request be merged if the changes are reverted back to original implementation I did.

**Comment by Sohaib-Ahmed21** [comment]
Closed as abandoned.
